content/client-lib-development-guide/features.textile

```diff
@@ -81,9 +81,11 @@ h3(#restclient). RestClient
 *** @(RSC7b1)@ When it is not possible to send the @X-Ably-Lib@ header, such as for @JSONP@ requests, the library version should be sent as a query param such as @lib=js-1.0.0@
 ** @(RSC7c)@ If the @addRequestIds@ client option is enabled, every REST request to Ably should include in a @request_id@ query string parameter a random string obtained by url-safe base64-encoding a sequence of at least 9 bytes obtained from a source of randomness. This request ID must remain the same if a request is retried to a fallback host per @RSC15@. Any log messages associated with the request should include the request ID. If the request fails, the request ID must be included in the @ErrorInfo@ returned to the user.
 ** @(RSC7d)@ An @Agent@ library identifier is used to identify the library type and version, as well as relevant information describing any underlying layers and the client platform.
-*** @(RSC7d1)@ The @Agent@ library identifier is composed of a series of @key[/value]@ entries joined by spaces where @key@ is the name of a product and @value@ is an optional version of the associated product which must be a valid "semantic version":https://semver.org/. This must include at least the @<library name>/<library version>@ entry followed optionally, in dependency order, the @name[/version]@ entry for underlying library layer or platform layers. An example would be @ably-flutter/1.2.0 ably-java/1.2.1 android/24@. The documentation for each library must define the content and meaning of the @Agent@ identifiers advertised by that library.
+*** @(RSC7d1)@ The @Agent@ library identifier is composed of a series of @key[/value]@ entries joined by spaces where @key@ is the name of a product and @value@ is an optional version of the associated product. This must include at least the @<library name>/<library version>@ entry followed optionally, in dependency order, the @name[/version]@ entry for underlying library layer or platform layers. An example would be @ably-flutter/1.2.0 ably-java/1.2.1 android/24@.
 *** @(RSC7d2)@ Where possible, the @Agent@ library identifier should be included in all HTTP and WebSocket requests to the Ably endpoint via an @Ably-Agent@ HTTP header.
 *** @(RSC7d3)@ Where the library is unable to specify a header when making a request (such as for @JSONP@ requests), the @Agent@ library identifier should be sent via an @agent@ query param whose value is the URI-encoded @Agent@ library identifier.
+*** @(RSC7d4)@ Where possible, product versions should be represented as a valid "semantic version":https://semver.org/. This is best-effort as we recognise that strict SemVer is not always available, or otherwise possible to derive from information available at runtime, especially in the case of products that represent operating system versions.
```

Comment:
This will require an equivalent relaxation of the server-side parsing of the Ably-Agent header, which currently expects a semantic version.

We could completely remove the parsing of the version and allow any string of non-whitespace characters, but this was previously problematic since absolute junk was appearing in our internal metrics.

What do you think about adding a `versionFormat` field to each agent entry in agents.json, and mapping those formats to regular expressions which are used for server-side parsing to avoid the junk? For example:

```
"agents": [
    {
        "identifier": "ably-java",
        "versioned": true,
        "emitters": [
            "ably-java"
        ],
        "type": "sdk",
        "versionFormat": "semver"
    },
    {
        "identifier": "android",
        "versioned": true,
        "emitters": [
            "ably-java",
            "ably-dotnet"
        ],
        "type": "os",
        "versionFormat": "integer"
    }
],
"versionFormats": {
    "semver": "<semver regex>",
    "integer": "\d+"
},
```

Reply:
The problem we're facing at the moment is that I think we'll struggle to define precise formats for every version we'll receive from some products (e.g. OS versions, custom Java runtimes, etc..) - as in, perhaps by definition, some level of "junk" will arrive.

Now that we've pinned down the product components, I think the problem will be much more manageable. My preference would be to remove the parsing of versions server-side for the moment, for our first iteration of this feature.

If it really becomes an issue down the line then perhaps we can look to add more filtering at the server-side. At least that will be managed in the one place, without need for complex client-side logic for every single client library.

Reply:
Ok, I'm happy to just remove the parsing and then re-consider if we see any issues re-surface 👍